python/ray/serve/kv_store.py

```diff
@@ -56,3 +60,143 @@ def delete(self, key):
         if not isinstance(key, str):
             raise TypeError("key must be a string, got: {}.".format(type(key)))
         return ray_kv._internal_kv_del(self._format_key(key))
+
+
+class RayExternalKVStore:
+    """Persistent version of RayInternalKVStore for cluster fault
+    tolerance. Writes to local disk in local / test mode, also external
+    storage (such as amazon S3) with provided path.
+
+    Supports string keys and bytes values, caller must handle serialization.
+    """
+
+    def __init__(
+            self,
+            namepsace: str,
+            bucket="",
+            s3_path="",
+            region_name="us-west-2",
+            aws_access_key_id=None,
+            aws_secret_access_key=None,
+            aws_session_token=None,
+            local_mode=True,
+    ):
+        self._namespace = namepsace
+        self._bucket = bucket
+        self._s3_path = s3_path
+        self._local_mode = local_mode
+        self._tombstone = b"\##DELETED\##"
+
+        self._s3 = boto3.client(
+            "s3",
+            region_name=region_name,
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+            aws_session_token=aws_session_token)
+
+    def get_key_with_namespace(self, key):
+        return "{ns}-{key}".format(ns=self._namespace, key=key)
+
+    def put(self, key, val):
+        """Put the key-value pair into the store.
+
+        Args:
+            key (str)
+            val (bytes)
+        """
+        if not isinstance(key, str):
+            raise TypeError("key must be a string, got: {}.".format(type(key)))
+        if not isinstance(val, bytes):
+            raise TypeError("val must be bytes, got: {}.".format(type(val)))
+
+        if self._local_mode:
+            with open("/tmp/ray_serve_checkpoint_key.txt",
+                      "a+") as key_file, open(
+                          "/tmp/ray_serve_checkpoint_val.txt",
+                          "ab+") as val_file:
+                key_file.write(self.get_key_with_namespace(key) + "\n")
+                val_file.write(val)
+                val_file.write(b"\n")
```

Comment:
Can you explain the file layout and semantics for this in a comment? This should be the docstring for the `LocalDiscKVStore`